### PR TITLE
convert npm dependencies to peerDependencies

### LINKED ISSUE #5560: declare peer dependencies as peer dependencies
Our package.json lists many peer dependencies as just dependencies, but should be reflected as peer dependencies:
 * reflect-metadata
 * rxjs
 * zone.js
 * es6-shim

This is to make sure duplicates don't end up in bundles, and to make sure Angular is sharing the same instances of modules. This is particularly important with RxJS where importing operators has a side effect of monkey-patching the `Observable` imported from `rxjs/Observable`.

### PR DESCRIPTION
**Pull request by IgorMinar**
Closes #5560

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,7 +30,8 @@
     "postinstall": "node tools/analytics/build-analytics success install npm-install-net && node tools/analytics/build-analytics start install npm-postinstall && node tools/npm/copy-npm-shrinkwrap && node tools/chromedriverpatch.js && webdriver-manager update && bower install && gulp pubget.dart && tsd reinstall --overwrite --clean --config modules/angular2/tsd.json && tsd reinstall --overwrite --clean --config tools/tsd.json && tsd reinstall --overwrite --config modules/angular1_router/tsd.json && node tools/analytics/build-analytics success install npm-postinstall && node tools/analytics/build-analytics success install npm-install",
     "test": "gulp test.all.js && gulp test.all.dart"
   },
-  "dependencies": {
+  "peerDependencies": {
+    "es6-shim": "^0.33.3",
     "reflect-metadata": "0.1.2",
     "rxjs": "^5.0.0-alpha.11",
```
Comment (pkozlowski-opensource) [inline package.json]: Shouldn't we be more strict about rxjs? If I recall properly alpha.12 was kind of broken and @robwormald had to go back to .11. In any case given the amount of changes in both libraries we might be more strict here.

**Comment by pkozlowski-opensource** [comment]
Travis build is not a flake, I think that we read from package.json dependencies in some of our scripts

File: package.json (hunk 1 above)
Comment (IgorMinar) [inline package.json]: I'll do that in a separate commit within this PR. do we know what version of RXJS is the good one?

**Comment by IgorMinar** [comment]
yay for CI. I'll fix the breakage.

**Comment by IgorMinar** [comment]
@pkozlowski-opensource should be all fixed now. I also added es6-promise as a peer dependency (it comes from zone.js). waiting for CI now..

**Comment by IgorMinar** [comment]
oh.. I still need to lock rxjs.. one more commit coming.

**Comment by IgorMinar** [comment]
done

File: package.json [hunk 2]
```diff
@@ -30,9 +30,11 @@
     "postinstall": "node tools/analytics/build-analytics success install npm-install-net && node tools/analytics/build-analytics start install npm-postinstall && node tools/npm/copy-npm-shrinkwrap && node tools/chromedriverpatch.js && webdriver-manager update && bower install && gulp pubget.dart && tsd reinstall --overwrite --clean --config modules/angular2/tsd.json && tsd reinstall --overwrite --clean --config tools/tsd.json && tsd reinstall --overwrite --config modules/angular1_router/tsd.json && node tools/analytics/build-analytics success install npm-postinstall && node tools/analytics/build-analytics success install npm-install",
     "test": "gulp test.all.js && gulp test.all.dart"
   },
-  "dependencies": {
+  "peerDependencies": {
+    "es6-promise": "^3.0.2",
```
Comment (jeffbcross) [inline package.json]: Why do we need this now?

Comment (IgorMinar) [inline package.json]: We always needed it. It's a transitive dep coming from zone

File: tools/broccoli/trees/node_tree.ts [hunk 3]
```diff
@@ -78,7 +78,7 @@ module.exports = function makeNodeTree(projects, destinationPath) {
     license: BASE_PACKAGE_JSON.license,
     repository: BASE_PACKAGE_JSON.repository,
     contributors: BASE_PACKAGE_JSON.contributors,
-    dependencies: BASE_PACKAGE_JSON.dependencies,
+    peerDependencies: BASE_PACKAGE_JSON.peerDependencies,
```
Comment (pkozlowski-opensource) [inline tools/broccoli/trees/node_tree.ts]: How it is going to work with npm3 where peer deps are not installed? The thing is that we still need all the runtime dependencies (zone etc.) to run tests, build bundles etc. So we still going to need to add them to devDependencies or dependencies.

**Comment by pkozlowski-opensource** [comment]
@IgorMinar the more I look at this PR the more I think that we are focusing on the wrong package.json. What we want to do is to have peer dependencies in the package.json of a module that we publish to npm.

Here we still want things as regular (dependencies or devDependencies) deps since we need zone and company to run tests, prepare bundles etc. 

What I'm saying is that we don't want to move zone & co to peereDependencies in this repo's package.json but do this only in package.json that we push to npm.

File: package.json (hunk 2 above)
Comment (pkozlowski-opensource) [inline package.json]: Isn't it part of zone.js build? Or are you saying that we need to have it for some other reason (typings)?

File: package.json [hunk 4]
```diff
@@ -30,9 +30,11 @@
...
+    "es6-shim": "^0.33.3",
```
Comment (pkozlowski-opensource) [inline package.json]: Do we want to force `es6-shim`? What about people that do projects for modern browsers only? Or want to use a different shim?

**Comment by jeffbcross** [comment]
@pkozlowski-opensource: @IgorMinar said you are right (as always) and that he is going to update the PR to keep `devDependencies` in our own package, `peerDependencies` in published packages.

**Comment by IgorMinar** [comment]
@pkozlowski-opensource you are right. my bad. I'm fixing that now.

**Comment by IgorMinar** [comment]
ci still failing, looking into that too

**Comment by IgorMinar** [comment]
changes made. running CI. if green then I need to squash commits and we can land this

**Comment by pkozlowski-opensource** [comment]
@IgorMinar what about my comments regarding `es6-promise` and `es6-shim`? 

I'm not sure why we need `es6-promise` (it is bundled with zone.js) and `es6-shim` is debatable IMO (what about people using recent browsers? Or ones using a different shim?)

File: modules/angular2/src/http/package.json [hunk 5]
```diff
@@ -7,10 +7,8 @@
...
   "license": "<%= packageJson.license %>",
   "repository": <%= JSON.stringify(packageJson.repository) %>,
-  "dependencies": {
-    "angular2": "<%= packageJson.version %>",
-    "rxjs": "<%= packageJson.dependencies['rxjs'] %>",
-    "reflect-metadata": "<%= packageJson.dependencies['reflect-metadata'] %>"
-  },
   "devDependencies": <%= JSON.stringify(packageJson.defaultDevDependencies) %>
+  "peerDependencies": {
+    "angular2": "<%= packageJson.version %>"
```
Comment (jeffbcross) [inline modules/angular2/src/http/package.json]: Should have a dependency on rxjs

Comment (jeffbcross) [inline modules/angular2/src/http/package.json]: I guess rxjs is already declared as a peerDependency of angular2, so as long as this package depends on angular2 rxjs isn't necessary.

Comment (jeffbcross) [inline modules/angular2/src/http/package.json]: And http isn't even published as a standalone package yet.

Comment (IgorMinar) [inline modules/angular2/src/http/package.json]: Jeff, rxjs comes in via angular. But I'll make it an explicit peer dep.

Pawel, see my comment about optionalPeerDependencies in one of my commit
messages. For now apps that don't need shims will still need to declare
them in their package.json but not use them.

**Comment by IgorMinar** [comment]
CI is green! I'm squashing the commits and applying merge label

**Comment by IgorMinar** [comment]
everything is resolved and rebased. good to be merged now

**Comment by jelbourn** [comment]
@IgorMinar Travis failures still look legitimate to me.

**Comment by IgorMinar** [comment]
looking into it

**Comment by IgorMinar** [comment]
sorry.. it was a bad squash. fixed now. rerunning CI

**Comment by jelbourn** [comment]
@mary-poppins isn't picking this one up, merging manually.

**Pull request closed without merging**